examples/modular-drivers/src/bme280_ulp.rs

```diff
@@ -0,0 +1,970 @@
+    dig_p7: i16,
+    dig_p8: i16,
+    dig_p9: i16,
+    dig_h1: u8,
+    dig_h2: i16,
+    dig_h3: u8,
+    dig_h4: i16,
+    dig_h5: i16,
+    dig_h6: i8,
+    t_fine: i32,
+}
+
+impl From<[u8; BME280_CALIB_LEN]> for CalibrationData {
+    fn from(data: [u8; BME280_CALIB_LEN]) -> Self {
+        Self {
+            dig_t1: u16::from_le_bytes([data[0], data[1]]),
+            dig_t2: i16::from_le_bytes([data[2], data[3]]),
+            dig_t3: i16::from_le_bytes([data[4], data[5]]),
+            dig_p1: u16::from_le_bytes([data[6], data[7]]),
+            dig_p2: i16::from_le_bytes([data[8], data[9]]),
+            dig_p3: i16::from_le_bytes([data[10], data[11]]),
+            dig_p4: i16::from_le_bytes([data[12], data[13]]),
+            dig_p5: i16::from_le_bytes([data[14], data[15]]),
+            dig_p6: i16::from_le_bytes([data[16], data[17]]),
+            dig_p7: i16::from_le_bytes([data[18], data[19]]),
+            dig_p8: i16::from_le_bytes([data[20], data[21]]),
+            dig_p9: i16::from_le_bytes([data[22], data[23]]),
+            dig_h1: data[25],
+            dig_h2: i16::from_le_bytes([data[26], data[27]]),
+            dig_h3: data[28],
+            dig_h4: (i16::from(data[29] << 4) | i16::from(data[30]) & 0xf),
+            dig_h5: ((i16::from(data[30]) & 0xf0) >> 4) | (i16::from(data[31]) << 4),
+            dig_h6: data[32] as i8,
+            t_fine: 0,
+        }
+    }
+}
+
+#[derive(Default, Debug, Clone)]
+struct Configuration {
+    filter: Filter,
+    smp_temp: OverSampling,
+    smp_hum: OverSampling,
+    smp_pres: OverSampling,
+    t_sb: Standby,
+    mode: Mode,
+}
+
+impl Configuration {
+    fn ctrl_meas_value(&self) -> u8 {
+        (self.smp_temp << 5) | (self.smp_pres << 2) | Into::<u8>::into(self.mode)
+    }
+    fn ctrl_hum_value(&self) -> u8 {
+        self.smp_hum.into()
+    }
+    fn ctrl_config_value(&self) -> u8 {
+        (self.t_sb << 5) | (self.filter << 2)
+    }
+}
+#[allow(dead_code)]
+struct Status {
+    measuring: u8,
+    im_update: u8,
+}
+
+impl From<u8> for Status {
+    fn from(value: u8) -> Self {
+        Self {
+            im_update: value & 0x1,
+            measuring: (value >> 3) & 0x1,
+        }
+    }
+}
+
+const BME280_DEV_ADDR: u8 = 0x76;
+const BME280_CHIP_ID: u8 = 0x60;
+const BME280_RESET: u8 = 0xB6;
+const BME280_CALIB_LEN: usize = 33;
+const BME280_CALIB_PART1: usize = 26;
+const BME280_MEAS_LEN: usize = 8;
+const BME280_INACTIVE_MEAS: u32 = 0x80000;
+
+#[derive(DoCommand)]
+struct BME280 {
+    config: Configuration,
+    driver: I2cHandleType,
+    calib: CalibrationData,
+    ulp: Option<ULP>,
+}
+
+impl Drop for BME280 {
+    fn drop(&mut self) {
+        if let Some(ulp) = &mut self.ulp {
+            if let Err(err) = ulp.start_ulp(self.config.clone()) {
+                log::error!("couldn't start BME280 ULP program reason : {}", err);
+            } else {
+                log::info!("started BME280 ULP program")
+            }
+        }
+    }
+}
```

Comment:
what prompts the dropping of the sensor?

Reply:
dropping the robot